File: 2016/day1.py
def part1():
    f = open('inputs/day1.txt')
    x = 0
    y = 0

    '''
    0 == N
    1 == N+1 == S-1 == E
    2 == S
    3 == W
    '''
    facing = 0

    for line in f:
        for instruction in line.split(","):
            instruction = instruction.strip()
            direction = instruction[0]
            distance = int(instruction[1:])

            if direction == "R":
                facing = (facing + 1) % 4
            else:
                facing = (facing -1) % 4


            if facing == 0:
                y += distance
            elif facing == 1:
                x += distance
            elif facing == 2:
                y -= distance
            elif facing == 3:
                x -= distance

    print "(part1): distance from start: ", abs(x) + abs(y)

def part2():
    visited = set()
    visited.add((0,0))

    current_set_size = len(visited)

    f = open('inputs/day1.txt')
    x = 0
    y = 0

    '''
    0 == N
    1 == N+1 == S-1 == E
    2 == S
    3 == W
    '''
    facing = 0

    found = False

    for line in f:
        for instruction in line.split(","):
            instruction = instruction.strip()
            direction = instruction[0]
            distance = int(instruction[1:])

            if direction == "R":
                facing = (facing + 1) % 4
            else:
                facing = (facing -1) % 4

            current = (x, y)
            print current
            final = None

            if facing == 0:
                final = (x, y+distance)
            elif facing == 1:
                final = (x+distance, y)
            elif facing == 2:
                final = (x, y-distance)
            elif facing == 3:
                final = (x+distance, y)

            for i in range(1, distance+1):
                if facing == 0:
                    current = (current[0], current[1]+1)
                elif facing == 1:
                    current = (current[0]+1, current[1])
                elif facing == 2:
                    current = (current[0], current[1]-1)
                elif facing == 3:
                    current = (current[0]-1, current[1])

                print current, visited


                if current in visited:
                    print "part2: ", current, abs(current[0]) + abs(current[1])
                    found = True
                    break
                else:
                    visited.add(current)

            if found:
                break


            x = current[0]
            y = current[1]
#part1()
part2()

File: 2016/go/day1/day1.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"regexp"
	"strconv"
	"strings"
)

type Pos struct {
	x   int
	y   int
	dir int
}

func part1(file string) {
	fmt.Println("Part1:")
	pos := Pos{0, 0, 0}

	r, _ := regexp.Compile("(R|L)(\\d+)")

	for _, cmd := range strings.Split(file, ", ") {
		reg := r.FindStringSubmatch(cmd)
		turn := reg[1]
		dist, _ := strconv.Atoi(reg[2])

		if turn == "R" {
			pos.dir += 1
		} else {
			pos.dir -= 1
		}

		pos.dir = (pos.dir) % 4
		if pos.dir < 0 {
			pos.dir += 4
		}

		mod := 0
		if pos.dir < 2 {
			mod = 1
		} else {
			mod = -1
		}

		if pos.dir%2 == 0 {
			pos.y += mod * dist
		} else {
			pos.x += mod * dist
		}
	}
	fmt.Println("Final Location: ", pos)
	fmt.Println("Blocks: ", math.Abs(float64(pos.x))+math.Abs(float64(pos.y)))
}

func part2(file string) {
	fmt.Println("Part2:")
	visited := make(map[string]bool)
	visited["(0,0)"] = true

	pos := Pos{0, 0, 0}

	r, _ := regexp.Compile("(R|L)(\\d+)")

	for _, cmd := range strings.Split(file, ", ") {
		reg := r.FindStringSubmatch(cmd)
		turn := reg[1]
		dist, _ := strconv.Atoi(reg[2])

		if turn == "R" {
			pos.dir += 1
		} else {
			pos.dir -= 1
		}

		pos.dir = (pos.dir) % 4
		if pos.dir < 0 {
			pos.dir += 4
		}

		mod := 0
		if pos.dir < 2 {
			mod = 1
		} else {
			mod = -1
		}

		done := false
		visits := travel(pos, dist, mod)
		fmt.Println("############")
		fmt.Println("current: ", pos)
		fmt.Println("visited: ", visited)
		fmt.Println("visits: ", visits)
		fmt.Println("############")
		for _, v := range visits {
			if visited[v] {
				fmt.Printf("Visited %s twice!\n", v)
				done = true
				break
			} else {
				visited[v] = true
			}
		}
		if done {
			break
		}
		pos = NewPos(pos, dist, mod)
	}

}

func travel(pos Pos, dist, mod int) (visited []string) {
	newPos := NewPos(pos, dist, mod)

	if pos.dir%2 == 0 {
		for i := pos.y + mod; i < newPos.y+mod; i += mod {
			pos.y = i
			visited = append(visited, fmt.Sprintf("(%d,%d)", pos.x, pos.y))
		}
	} else {
		for i := pos.x + mod; i < newPos.x+mod; i += mod {
			pos.x = i
			visited = append(visited, fmt.Sprintf("(%d,%d)", pos.x, pos.y))
		}
	}

	return visited
}

func NewPos(pos Pos, dist, mod int) Pos {
	newPos := Pos{pos.x, pos.y, pos.dir}
	if pos.dir%2 == 0 {
		newPos.y += mod * dist
	} else {
		newPos.x += mod * dist
	}
	return newPos
}

func main() {
	dat, _ := ioutil.ReadFile("input.txt")
	file := string(dat)

	part1(file)
	part2(file)
}
